### PR TITLE
Fixed inefficient Stream API call chains ending with count()

### PR DESCRIPTION
**Pull request by dk2k**
Fixed inefficient Stream API call chains ending with count()

### CODE REVIEW

File: modules/percolator/src/main/java/org/opensearch/percolator/QueryAnalyzer.java
Comment (dk2k) [reply modules/percolator/src/main/java/org/opensearch/percolator/QueryAnalyzer.java:516]: @dbwiddis 
I think there is a mistake in the proposed code for single iteration.
Negation for condition 
extractions.stream().anyMatch(queryExtraction -> queryExtraction.range != null)
is
extractions.stream().allMatch(queryExtraction -> queryExtraction.range == null)

And hence you can't combine noneMatch() with allMatch() into single noneMatch()

Comment (dk2k) [reply modules/percolator/src/main/java/org/opensearch/percolator/QueryAnalyzer.java:516]: @dbwiddis
I think there is a logical mistake in the proposed code for single iteration.
Negation for condition
// cond1
extractions.stream().anyMatch(queryExtraction -> queryExtraction.range != null) - predicate A
isn't
// cond2
extractions.stream().noneMatch(queryExtraction -> queryExtraction.range == null) - predicate !A

noneMatch() is opposite for allMatch(), not for anyMatch()

Say, we have 3 elements in collection extractions and only one of them has non-null range. cond1 will be true, cond2 will be false

And hence there is no ground for combining into single noneMatch().

Comment (sandeshkr419) [reply modules/percolator/src/main/java/org/opensearch/percolator/QueryAnalyzer.java:516]: I think using a `for` loop might be simpler and cleaner approach here. 

You could refactor the entire method in a single for loop:

```
private static int minTermLength(Set<QueryExtraction> extractions) {
    boolean hasTerm = false;
    boolean hasRange = false;
    int min = Integer.MAX_VALUE;

    for (QueryExtraction qt : extractions) {
        if (qt.term != null) {
            hasTerm = true;
            min = Math.min(min, qt.bytes().length);
        }
        if (qt.range != null) {
            hasRange = true;
        }
    }

    // If there are no terms but there are ranges, return Integer.MIN_VALUE
    if (!hasTerm && hasRange) {
        return Integer.MIN_VALUE;
    }

    return min;
}
```

We should gain some performance gains by doing this for sure.

Comment (dbwiddis) [reply modules/percolator/src/main/java/org/opensearch/percolator/QueryAnalyzer.java:516]: > @dbwiddis I think there is a logical mistake in the proposed code for single iteration. Negation for condition // cond1 extractions.stream().anyMatch(queryExtraction -> queryExtraction.range != null) - predicate A isn't // cond2 extractions.stream().noneMatch(queryExtraction -> queryExtraction.range == null) - predicate !A

You're correct, I made a logic error.  Still, the existing implementation takes more than one iteration to achieve success.

Your existing code does this:
1. Iterate all the extractions and verify none of the `.term` fields are non-null. 
2. If we've done a complete iteration and all the `.term` fields are non-null, iterate them again to find at least one `.range` that is non-null.

Here are the possible outcomes of this logic:
1. Any single .term non-null, fail fast (takes 0.5 iterations on average)
2. All .term are null, at least one .range not null = success (takes 1.5 iterations on average)
3. All .term are null, all .range are null = failure (takes 2 full iterations)

As @sandeshkr419 said this is probably more clear in a loop that we iterate once, failing fast on condition 1 (same as existing) and setting a boolean when a non-null .range is found.  For both conditions 2 and 3 we will do exactly one iteration to verify null .term, and then just check the value of the boolean.

Comment (dk2k) [reply modules/percolator/src/main/java/org/opensearch/percolator/QueryAnalyzer.java]: @sandeshkr419  thanks, I applied your method

Comment (sandeshkr419) [reply modules/percolator/src/main/java/org/opensearch/percolator/QueryAnalyzer.java]: This looks good. 
I'll still wait for the CI to finish to see the codecov report as well before approving the change. We'd ideally want to increase the code coverage and span coverage across any change in this code.

**Review by dbwiddis — APPROVED** [review]

**Comment by sandeshkr419** [comment]
```
* What went wrong:
Execution failed for task ':distribution:bwc:minor:buildBwcLinuxTar'.
> Building 2.17.0 didn't generate expected file /var/jenkins/workspace/gradle-check/search/distribution/bwc/minor/build/bwc/checkout-2.x/distribution/archives/linux-tar/build/distributions/opensearch-min-2.17.0-SNAPSHOT-linux-x64.tar.gz
```

@dk2k I think you will have to rebase your changes against main branch for gradle checks to succeed since 2.17 is released after this PR was opened up.

File: modules/percolator/src/main/java/org/opensearch/percolator/QueryAnalyzer.java
Comment (dbwiddis) [reply modules/percolator/src/main/java/org/opensearch/percolator/QueryAnalyzer.java]: Thanks @sandeshkr419 for the loop suggestion. The stream code wasn't very readable as I demonstrated by misunderstanding it multiple times. I really like this solution.

**Review by sandeshkr419 — APPROVED** [review]

**Review by dbwiddis — COMMENTED** [review]
Thanks for these changes, @dk2k !   Appreciate your patience as we've reviewed them.

One last thing we'd like is an entry in CHANGELOG.md summarizing this PR.  There's an open issue on percolator performance (#16285) so adding an entry showing an attempt to improve at least a portion of it would be helpful.

Ready to approve pending this entry.

**Review by dbwiddis — APPROVED** [review]
Re-approving, will merge once CI completes

**Review by dbwiddis — APPROVED** [review]

**Pull request merged**
